File: projects/group/legacy-flutter/app/lib/data/event_log.dart
import 'package:meta/meta.dart';
import 'package:riverpod/riverpod.dart';

import '../scenarios/events.dart';

@immutable
class EventLogEntry {
  const EventLogEntry({required this.event, this.riskScore, this.tags = const []});

  final ScamEvent event;
  final double? riskScore;
  final List<String> tags;

  EventLogEntry copyWith({double? riskScore, List<String>? tags}) {
    return EventLogEntry(
      event: event,
      riskScore: riskScore ?? this.riskScore,
      tags: tags ?? this.tags,
    );
  }
}

class EventLog extends Notifier<List<EventLogEntry>> {
  @override
  List<EventLogEntry> build() => const [];

  void add(ScamEvent event) {
    state = [...state, EventLogEntry(event: event)];
  }

  void annotate(String id, {double? risk, List<String>? tags}) {
    state = [
      for (final e in state)
        if (e.event.id == id)
          e.copyWith(riskScore: risk, tags: tags)
        else
          e,
    ];
  }

  void clear() {
    state = const [];
  }

  Iterable<ScamEvent> within(Duration window, {DateTime? now}) {
    final n = now ?? DateTime.now();
    final cutoff = n.subtract(window);
    return state
        .where((e) => e.event.timestamp.isAfter(cutoff))
        .map((e) => e.event);
  }
}

final eventLogProvider =
    NotifierProvider<EventLog, List<EventLogEntry>>(EventLog.new);

File: projects/group/streamlit_app/guardian/data/event_log.py
"""Append-only event log with per-event risk annotations.

1:1 port of ``app/lib/data/event_log.dart``. Lives in session state; not
thread-safe, but Streamlit reruns on the main thread so that's fine.
"""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from datetime import datetime, timedelta
from typing import Iterable, Iterator

from guardian.scenarios.events import ScamEvent


@dataclass
class EventLogEntry:
    event: ScamEvent
    risk_score: float | None = None
    tags: list[str] = field(default_factory=list)


class EventLog:
    def __init__(self) -> None:
        self._entries: list[EventLogEntry] = []

    def __iter__(self) -> Iterator[EventLogEntry]:
        return iter(self._entries)

    def __len__(self) -> int:
        return len(self._entries)

    @property
    def entries(self) -> list[EventLogEntry]:
        return list(self._entries)

    def add(self, event: ScamEvent) -> None:
        self._entries.append(EventLogEntry(event=event))

    def annotate(
        self,
        event_id: str,
        *,
        risk: float | None = None,
        tags: list[str] | None = None,
    ) -> None:
        for i, entry in enumerate(self._entries):
            if entry.event.id == event_id:
                self._entries[i] = replace(
                    entry,
                    risk_score=risk if risk is not None else entry.risk_score,
                    tags=tags if tags is not None else entry.tags,
                )
                return

    def clear(self) -> None:
        self._entries.clear()

    def within(self, window: timedelta, now: datetime | None = None) -> Iterable[ScamEvent]:
        n = now or datetime.now()
        cutoff = n - window
        for entry in self._entries:
            if entry.event.timestamp > cutoff:
                yield entry.event
